File: frontend/src/lib/api/country.ts
import { apiClient } from "./apiClient";

export type ApiCountry = {
  id: number;
  name: string;
  continent_id: number;
  continent_name?: string;
  flag_image_url?: string;
  map_image_url?: string;
  capital?: string;
  population?: number;
  area_km2?: number;
  language?: string;
  description?: string;
  is_featured?: number;
  currency?: string;
  area_comparison?: string;
  population_comparison?: string;
  greeting?: string;
  overview?: string;
  map_note?: string;
  clothing?: string;
  food?: string;
  house?: string;
  color?: string;
};

export type ApiContinent = {
  id: number;
  name: string;
  color: string;
  population?: number;
  area_km2?: number;
  country_count?: number;
};

export async function getContinents() {
  const { data } = await apiClient.get<ApiContinent[]>("/api/continents");
  return data;
}

export async function getCountries() {
  const { data } = await apiClient.get<ApiCountry[]>("/api/countries");
  return data;
}

export async function getFeaturedCountries() {
  const { data } = await apiClient.get<ApiCountry[]>("/api/countries/featured");
  return data;
}

export async function getCountry(id: number) {
  const { data } = await apiClient.get<ApiCountry>(`/api/countries/${id}`);
  return data;
}

let countryCache: ApiCountry[] | null = null;

export async function getCountryByName(name: string) {
  if (!countryCache) {
    countryCache = await getCountries();
  }
  return countryCache.find((country) => country.name === name) ?? null;
}

export async function getCountryIdByName(name: string) {
  const country = await getCountryByName(name);
  return country?.id ?? null;
}

File: backend/src/main/java/com/baeum/entity/Country.java
package com.baeum.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@Table(name = "countries")
@Getter
@NoArgsConstructor
public class Country {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String name;

    @Column(name = "continent_id")
    private Long continentId;

    @Column(name = "flag_image_url")
    private String flagImageUrl;

    @Column(name = "map_image_url")
    private String mapImageUrl;

    private String capital;
    private Long population;

    @Column(name = "area_km2")
    private Long areaKm2;

    private String language;
    private String description;

    @Column(name = "is_featured")
    private Integer isFeatured;
}
